File: scripts/filter_lichess_games.py
import argparse
import io
import subprocess
import sys
from pathlib import Path

import chess.pgn

# Examples:
# uv run python scripts/filter_lichess_games.py \
#   --input data/lichess_open_database/lichess_db_standard_rated_2024-01.pgn.zst \
#   --max-games 10000
#
# uv run python scripts/filter_lichess_games.py \
#   --input data/lichess_open_database/lichess_db_standard_rated_2024-01.pgn.zst \
#   --min-elo 1800 --min-speed blitz --variant Standard
#
# uv run python scripts/filter_lichess_games.py \
#   --input data/lichess_open_database/lichess_db_standard_rated_2024-01.pgn \
#   --result 1-0 --opening-contains "Sicilian" --max-games 50000


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Filter Lichess PGN games and write matching games to a new PGN file."
    )

    parser.add_argument("--input", required=True, help="Input .pgn or .pgn.zst file")
    parser.add_argument(
        "--out-dir",
        default="data/filtered_lichess",
        help="Directory to store filtered PGN output.",
    )
    parser.add_argument(
        "--output-name",
        default=None,
        help="Optional output filename. Default: <input_stem>.filtered.pgn",
    )
    parser.add_argument(
        "--max-games",
        type=int,
        default=None,
        help="Stop after writing this many matched games.",
    )

    parser.add_argument("--min-elo", type=int, default=None)
    parser.add_argument("--max-elo", type=int, default=None)
    parser.add_argument("--time-control-prefix", default=None)
    parser.add_argument(
        "--min-speed",
        choices=["bullet", "blitz", "rapid", "classical"],
        default=None,
        help="Keep games at this speed or slower (e.g. blitz keeps blitz/rapid/classical).",
    )
    parser.add_argument("--variant", default=None)
    parser.add_argument("--result", choices=["1-0", "0-1", "1/2-1/2"], default=None)
    parser.add_argument("--opening-contains", default=None)

    args = parser.parse_args()

    if args.min_elo is not None and args.max_elo is not None and args.min_elo > args.max_elo:
        parser.error("--min-elo must be <= --max-elo")

    if args.max_games is not None and args.max_games < 1:
        parser.error("--max-games must be >= 1")

    return args


def open_input_text_stream(path: Path):
    if path.suffix == ".zst":
        proc = subprocess.Popen(
            ["zstd", "-dc", str(path)],
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
        )
        assert proc.stdout is not None
        text_stream = io.TextIOWrapper(proc.stdout, encoding="utf-8", errors="replace")
        return text_stream, proc

    text_stream = path.open("r", encoding="utf-8", errors="replace")
    return text_stream, None


def parse_elo(headers: chess.pgn.Headers, key: str) -> int | None:
    value = headers.get(key)
    if not value:
        return None
    try:
        return int(value)
    except ValueError:
        return None


def parse_time_control_seconds(value: str) -> tuple[int, int] | None:
    # Expected format is "<base>+<increment>", for example "180+2".
    if "+" not in value:
        return None
    left, right = value.split("+", 1)
    try:
        return int(left), int(right)
    except ValueError:
        return None


def speed_rank(headers: chess.pgn.Headers) -> int | None:
    tc = headers.get("TimeControl", "")
    parsed = parse_time_control_seconds(tc)
    if parsed is None:
        return None

    base_seconds, increment_seconds = parsed
    estimated_seconds = base_seconds + (40 * increment_seconds)

    # Lichess-style speed buckets based on estimated game duration.
    # bullet: <3m, blitz: 3m-<8m, rapid: 8m-<25m, classical: >=25m
    if estimated_seconds < 180:
        return 0
    if estimated_seconds < 480:
        return 1
    if estimated_seconds < 1500:
        return 2
    return 3


def matches_filters(game: chess.pgn.Game, args: argparse.Namespace) -> bool:
    headers = game.headers

    if args.variant and headers.get("Variant") != args.variant:
        return False

    if args.result and headers.get("Result") != args.result:
        return False

    if args.time_control_prefix:
        tc = headers.get("TimeControl", "")
        if not tc.startswith(args.time_control_prefix):
            return False

    if args.min_speed:
        rank = speed_rank(headers)
        if rank is None:
            return False
        required_rank = {"bullet": 0, "blitz": 1, "rapid": 2, "classical": 3}[args.min_speed]
        if rank < required_rank:
            return False

    if args.opening_contains:
        opening = headers.get("Opening", "")
        if args.opening_contains.lower() not in opening.lower():
            return False

    if args.min_elo is not None or args.max_elo is not None:
        white_elo = parse_elo(headers, "WhiteElo")
        black_elo = parse_elo(headers, "BlackElo")
        if white_elo is None or black_elo is None:
            return False
        min_player_elo = min(white_elo, black_elo)
        max_player_elo = max(white_elo, black_elo)

        if args.min_elo is not None and min_player_elo < args.min_elo:
            return False
        if args.max_elo is not None and max_player_elo > args.max_elo:
            return False

    return True


def default_output_name(input_path: Path) -> str:
    if input_path.name.endswith(".pgn.zst"):
        base = input_path.name[:-8]
    else:
        base = input_path.stem
    return f"{base}.filtered.pgn"


def main() -> int:
    args = parse_args()
    input_path = Path(args.input)
    if not input_path.exists():
        print(f"File not found: {input_path}", file=sys.stderr)
        return 1

    output_dir = Path(args.out_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    out_name = args.output_name or default_output_name(input_path)
    output_path = output_dir / out_name

    games_seen = 0
    games_written = 0

    in_stream, zstd_proc = open_input_text_stream(input_path)

    try:
        with output_path.open("w", encoding="utf-8") as out_f:
            while True:
                game = chess.pgn.read_game(in_stream)
                if game is None:
                    break

                games_seen += 1

                if not matches_filters(game, args):
                    continue

                print(game, file=out_f, end="\n\n")
                games_written += 1

                if args.max_games is not None and games_written >= args.max_games:
                    break

                if games_seen % 10000 == 0:
                    print(f"progress: seen={games_seen} written={games_written}")
    finally:
        in_stream.close()
        if zstd_proc is not None:
            zstd_proc.wait()

    print(f"done: seen={games_seen} written={games_written}")
    print(f"output: {output_path}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/filter_lichess_games.go
package main

import (
	"bufio"
	"errors"
	"flag"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

// Example usage:
// go run scripts/filter_lichess_games.go --input data/lichess_open_database/lichess_db_standard_rated_2026-04.pgn --max-games 10000
// go run scripts/filter_lichess_games.go --input data/lichess_open_database/lichess_db_standard_rated_2026-04.pgn --min-elo 1800 --min-speed blitz --variant Standard
// go run scripts/filter_lichess_games.go --input data/lichess_open_database/lichess_db_standard_rated_2026-04.pgn --time-control-prefix 600+ --test-run --max-games 5
// go run scripts/filter_lichess_games.go --input data/lichess_open_database/lichess_db_standard_rated_2026-04.pgn --rated-only --max-games 5000

type Speed int

const (
	Bullet Speed = iota
	Blitz
	Rapid
	Classical
)

type GameRecord struct {
	Raw     string
	Headers map[string]string
	Index   int
}

type FilterConfig struct {
	MinElo            *int
	MaxElo            *int
	TimeControlPrefix string
	MinSpeed          *Speed
	Variant           string
	RatedOnly         bool
}

type cliConfig struct {
	Input             string
	OutDir            string
	OutputName        string
	MaxGames          int
	HasMaxGames       bool
	TestRun           bool
	TimeControlPrefix string
	Variant           string
	RatedOnly         bool
	MinElo            *int
	MaxElo            *int
	MinSpeed          *Speed
}

func main() {
	start := time.Now()

	cfg, err := parseFlags()
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}

	if err := run(cfg, start); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
}

func parseFlags() (cliConfig, error) {
	var cfg cliConfig
	var minEloFlag int
	var maxEloFlag int
	var minSpeedFlag string

	flag.StringVar(&cfg.Input, "input", "", "Input .pgn file")
	flag.StringVar(&cfg.OutDir, "out-dir", "data/filtered_lichess", "Directory to store filtered PGN output")
	flag.StringVar(&cfg.OutputName, "output-name", "", "Optional output filename. Default: <input_stem>.filtered.pgn")
	flag.IntVar(&cfg.MaxGames, "max-games", 0, "Stop after writing this many matched games")
	flag.IntVar(&minEloFlag, "min-elo", 0, "Minimum allowed min(player Elo)")
	flag.IntVar(&maxEloFlag, "max-elo", 0, "Maximum allowed max(player Elo)")
	flag.StringVar(&cfg.TimeControlPrefix, "time-control-prefix", "", "Keep games where TimeControl starts with this value")
	flag.StringVar(&minSpeedFlag, "min-speed", "", "Minimum speed bucket: bullet|blitz|rapid|classical")
	flag.StringVar(&cfg.Variant, "variant", "", "Exact Variant header match")
	flag.BoolVar(&cfg.RatedOnly, "rated-only", false, "Keep only rated games")
	flag.BoolVar(&cfg.TestRun, "test-run", false, "Print parsed values and match outcome for each game")

	flag.Parse()

	if cfg.Input == "" {
		return cfg, errors.New("--input is required")
	}
	if !strings.HasSuffix(strings.ToLower(cfg.Input), ".pgn") {
		return cfg, errors.New("--input must be a .pgn file for v1")
	}

	if cfg.MaxGames < 0 {
		return cfg, errors.New("--max-games must be >= 1 when provided")
	}
	if cfg.MaxGames > 0 {
		cfg.HasMaxGames = true
	}

	if minEloWasSet() {
		if minEloFlag < 0 {
			return cfg, errors.New("--min-elo must be >= 0")
		}
		cfg.MinElo = &minEloFlag
	}
	if maxEloWasSet() {
		if maxEloFlag < 0 {
			return cfg, errors.New("--max-elo must be >= 0")
		}
		cfg.MaxElo = &maxEloFlag
	}
	if cfg.MinElo != nil && cfg.MaxElo != nil && *cfg.MinElo > *cfg.MaxElo {
		return cfg, errors.New("--min-elo must be <= --max-elo")
	}

	if minSpeedFlag != "" {
		speed, ok := parseSpeed(minSpeedFlag)
		if !ok {
			return cfg, errors.New("--min-speed must be one of: bullet, blitz, rapid, classical")
		}
		cfg.MinSpeed = &speed
	}

	return cfg, nil
}

func run(cfg cliConfig, start time.Time) error {
	inFile, err := os.Open(cfg.Input)
	if err != nil {
		return fmt.Errorf("failed to open input file: %w", err)
	}
	defer inFile.Close()

	if err := os.MkdirAll(cfg.OutDir, 0o755); err != nil {
		return fmt.Errorf("failed to create output dir: %w", err)
	}

	outName := cfg.OutputName
	if outName == "" {
		outName = defaultOutputName(cfg.Input)
	}
	outPath := filepath.Join(cfg.OutDir, outName)

	outFile, err := os.Create(outPath)
	if err != nil {
		return fmt.Errorf("failed to create output file: %w", err)
	}
	defer outFile.Close()

	filterCfg := FilterConfig{
		MinElo:            cfg.MinElo,
		MaxElo:            cfg.MaxElo,
		TimeControlPrefix: cfg.TimeControlPrefix,
		MinSpeed:          cfg.MinSpeed,
		Variant:           cfg.Variant,
		RatedOnly:         cfg.RatedOnly,
	}

	gamesSeen := 0
	gamesWritten := 0
	lastProgressLog := start

	for game := range parseGames(inFile) {
		gamesSeen++
		match, reasons := matches(game, filterCfg)

		if cfg.TestRun {
			printTestRun(game, match, reasons)
		}

		if time.Since(lastProgressLog) >= time.Minute {
			elapsed := time.Since(start)
			fmt.Printf(
				"progress: seen=%d written=%d elapsed=%s\n",
				gamesSeen,
				gamesWritten,
				elapsed.Round(time.Second),
			)
			lastProgressLog = time.Now()
		}

		if !match {
			continue
		}

		if _, err := outFile.WriteString(game.Raw); err != nil {
			return fmt.Errorf("failed writing matched game: %w", err)
		}
		if !strings.HasSuffix(game.Raw, "\n\n") {
			if _, err := outFile.WriteString("\n\n"); err != nil {
				return fmt.Errorf("failed writing game separator: %w", err)
			}
		}

		gamesWritten++
		if cfg.HasMaxGames && gamesWritten >= cfg.MaxGames {
			break
		}
	}

	elapsed := time.Since(start)
	fmt.Printf("done: seen=%d written=%d\n", gamesSeen, gamesWritten)
	fmt.Printf("output: %s\n", outPath)
	fmt.Printf("elapsed: %s\n", elapsed.Round(time.Millisecond))

	return nil
}

func parseGames(reader io.Reader) <-chan GameRecord {
	out := make(chan GameRecord)
	go func() {
		defer close(out)

		scanner := bufio.NewScanner(reader)
		const maxLine = 1024 * 1024
		buf := make([]byte, 0, 64*1024)
		scanner.Buffer(buf, maxLine)

		var rawBuilder strings.Builder
		headers := map[string]string{}
		inGame := false
		moveStarted := false
		gameIndex := 0

		emitGame := func() {
			if !inGame {
				return
			}
			gameIndex++
			raw := rawBuilder.String()
			copiedHeaders := make(map[string]string, len(headers))
			for k, v := range headers {
				copiedHeaders[k] = v
			}
			out <- GameRecord{Raw: raw, Headers: copiedHeaders, Index: gameIndex}
			rawBuilder.Reset()
			headers = map[string]string{}
			inGame = false
			moveStarted = false
		}

		for scanner.Scan() {
			line := scanner.Text()
			lineWithNewline := line + "\n"
			trimmed := strings.TrimSpace(line)

			if strings.HasPrefix(trimmed, "[") {
				if inGame && moveStarted {
					emitGame()
				}
				inGame = true
				rawBuilder.WriteString(lineWithNewline)
				if key, value, ok := parseHeaderLine(trimmed); ok {
					headers[key] = value
				}
				continue
			}

			if inGame {
				rawBuilder.WriteString(lineWithNewline)
				if trimmed != "" {
					moveStarted = true
				}
			}
		}

		if inGame {
			emitGame()
		}
	}()

	return out
}

func parseHeaderLine(line string) (string, string, bool) {
	if !strings.HasPrefix(line, "[") || !strings.HasSuffix(line, "]") {
		return "", "", false
	}
	inner := strings.TrimSuffix(strings.TrimPrefix(line, "["), "]")
	firstQuote := strings.Index(inner, "\"")
	lastQuote := strings.LastIndex(inner, "\"")
	if firstQuote <= 0 || lastQuote <= firstQuote {
		return "", "", false
	}
	key := strings.TrimSpace(inner[:firstQuote])
	value := inner[firstQuote+1 : lastQuote]
	if key == "" {
		return "", "", false
	}
	return key, value, true
}

func matches(game GameRecord, cfg FilterConfig) (bool, []string) {
	reasons := make([]string, 0)
	headers := game.Headers

	if cfg.Variant != "" && headers["Variant"] != cfg.Variant {
		reasons = append(reasons, "variant_mismatch")
	}

	if cfg.TimeControlPrefix != "" {
		tc := headers["TimeControl"]
		if !strings.HasPrefix(tc, cfg.TimeControlPrefix) {
			reasons = append(reasons, "time_control_prefix_mismatch")
		}
	}

	if cfg.MinElo != nil || cfg.MaxElo != nil {
		whiteElo, errWhite := strconv.Atoi(headers["WhiteElo"])
		blackElo, errBlack := strconv.Atoi(headers["BlackElo"])
		if errWhite != nil || errBlack != nil {
			reasons = append(reasons, "elo_parse_fail")
		} else {
			minPlayerElo := minInt(whiteElo, blackElo)
			maxPlayerElo := maxInt(whiteElo, blackElo)

			if cfg.MinElo != nil && minPlayerElo < *cfg.MinElo {
				reasons = append(reasons, "min_elo_fail")
			}
			if cfg.MaxElo != nil && maxPlayerElo > *cfg.MaxElo {
				reasons = append(reasons, "max_elo_fail")
			}
		}
	}

	if cfg.MinSpeed != nil {
		tc := headers["TimeControl"]
		rank, ok := speedRankFromTimeControl(tc)
		if !ok {
			reasons = append(reasons, "min_speed_parse_fail")
		} else if rank < *cfg.MinSpeed {
			reasons = append(reasons, "min_speed_fail")
		}
	}

	if cfg.RatedOnly && !isRatedGame(headers) {
		reasons = append(reasons, "rated_only_fail")
	}

	return len(reasons) == 0, reasons
}

func isRatedGame(headers map[string]string) bool {
	ratedValue := strings.ToLower(strings.TrimSpace(headers["Rated"]))
	if ratedValue != "" {
		return ratedValue == "true" || ratedValue == "1" || ratedValue == "yes"
	}
	event := strings.TrimSpace(headers["Event"])
	return strings.Contains(strings.ToLower(event), "rated")
}

func speedRankFromTimeControl(tc string) (Speed, bool) {
	parts := strings.SplitN(tc, "+", 2)
	if len(parts) != 2 {
		return Bullet, false
	}
	base, errBase := strconv.Atoi(parts[0])
	inc, errInc := strconv.Atoi(parts[1])
	if errBase != nil || errInc != nil {
		return Bullet, false
	}
	estimatedSeconds := base + (40 * inc)
	if estimatedSeconds < 180 {
		return Bullet, true
	}
	if estimatedSeconds < 480 {
		return Blitz, true
	}
	if estimatedSeconds < 1500 {
		return Rapid, true
	}
	return Classical, true
}

func parseSpeed(s string) (Speed, bool) {
	switch strings.ToLower(strings.TrimSpace(s)) {
	case "bullet":
		return Bullet, true
	case "blitz":
		return Blitz, true
	case "rapid":
		return Rapid, true
	case "classical":
		return Classical, true
	default:
		return Bullet, false
	}
}

func printTestRun(game GameRecord, match bool, reasons []string) {
	headers := game.Headers
	status := "MATCH"
	if !match {
		status = "NO_MATCH"
	}

	reasonText := "-"
	if len(reasons) > 0 {
		reasonText = strings.Join(reasons, ",")
	}

	fmt.Printf(
		"game=%d whiteElo=%q blackElo=%q timeControl=%q variant=%q rated=%q result=%s reasons=%s\n",
		game.Index,
		headers["WhiteElo"],
		headers["BlackElo"],
		headers["TimeControl"],
		headers["Variant"],
		headers["Rated"],
		status,
		reasonText,
	)
}

func defaultOutputName(inputPath string) string {
	base := filepath.Base(inputPath)
	ext := filepath.Ext(base)
	if ext != "" {
		base = strings.TrimSuffix(base, ext)
	}
	return base + ".filtered.pgn"
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func minEloWasSet() bool {
	return flagWasSet("min-elo")
}

func maxEloWasSet() bool {
	return flagWasSet("max-elo")
}

func flagWasSet(name string) bool {
	set := false
	flag.Visit(func(f *flag.Flag) {
		if f.Name == name {
			set = true
		}
	})
	return set
}
